File: leetcode/321_Create_Maximum_Number.py
# @Author: Xingxing Huang
# @Date: 2017.04.02
# @Time: (m*n)*k
# 思路： 暴力解法，每次从前面找到最大的数加入到结果中。当剩余数组长度之和为给定长度，返回两个数组的merge。但是无法解决两个数组中前面几个数都包含最大得数时，选择哪个数组中的数的问题。
class Solution(object):
    def maxNumber(self, nums1, nums2, k):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :type k: int
        :rtype: List[int]
        """
        len1 = len(nums1)
        len2 = len(nums2)
        if len1 == 0:
            return self.select(nums2, k)
        if len2 == 0:
            return self.select(nums1, k)
        maxValue = -1
        index1 = -1
        index2 = -1
        for i, n1 in enumerate(nums1):
            for j,n2 in enumerate(nums2):
                if len1 + len2 - i - j < k:
                    continue
                if n1 > maxValue:
                    index1 = i
                    index2 = -1
                    maxValue = n1
                if n2 > maxValue:
                    index1 = -1
                    index2 = j
                    maxValue = n2
        if k == 1:
            return [maxValue]
        if index1 == -1:
            res = self.maxNumber(nums1, nums2[(index2 + 1):], k - 1) 
        else:
            res = self.maxNumber(nums1[(index1 + 1):], nums2, k - 1)
        return [maxValue] + res
        
    def select(self, nums, k):
        if len(nums) == k:
            return nums
        maxValue = nums[0]
        for i, n in enumerate(nums):
            if len(nums) - i < k:
                return [maxValue] + self.select(nums[i:], k - 1)
            if n > maxValue:
                maxValue = n
        return [maxValue]
            

File: leetcode/321_Create_Maximum_Number.java
// 10.04
public class Solution { 
    public int[] maxNumber(int[] nums1, int[] nums2, int k) {
        if (k == 0) return new int[0];
        int m = nums1.length;
        int n = nums2.length;
        if (m + n < k) return null;
        if (m + n == k) return merge(nums1, nums2, k);
        
        // 从两个数组中分别选出一定数目，组成k大小的数组。循环查看最满足条件的情况。第一个数组可以取元素数目从min到max
        int max = m >= k ? k : m;
        int min = n >= k ? 0 : k - n;
        int[] results = new int[k];
        for (int i = 0; i < results.length; i++) results[i] = Integer.MIN_VALUE;
        for (int i = min; i <= max; i++) {
            int[] temp = merge(getMax(nums1, i), getMax(nums2, k - i), k);
            results = isGreater(results, 0, temp, 0) ? results : temp;
        }
        return results;
    }
        private int[] merge(int[] nums1, int[] nums2, int k) {
        int[] results = new int[k];
        if (k == 0) return results;
        int i = 0, j = 0;
        for(int l = 0; l < k; ++l) {
            results[l] = isGreater(nums1, i, nums2, j) ? nums1[i++] : nums2[j++];
        }
        return results;
    }

    private boolean isGreater(int[] nums1, int i, int[] nums2, int j) {
        // 比较两个数的大小
        for(; i < nums1.length && j < nums2.length; ++i, ++j) {
            if (nums1[i] > nums2[j])
                return true;
            if (nums1[i] < nums2[j])
                return false;
        }
        return i != nums1.length;
    }

    private int[] getMax(int[] nums, int k) {
        if (k == 0)
            return new int[0];
        int[] results = new int[k];
        int i = 0;
        for(int j = 0; j < nums.length; ++j) {
            // 如果数组后面还有充足的元素，并且选出的元素个数不为0。新的元素比当前元素大。我们直接将数组里面所有比当前元素小的元素剔除
            while(nums.length - j + i > k && i > 0 && results[i-1] < nums[j])
                i--;
            // 还能添加元素 就继续添加。
            if (i < k)
                results[i++] = nums[j];
        }
        return results;
    }
}
// // http://www.jiuzhang.com/solutions/create-maximum-number/
// public class Solution {
//     public int[] maxNumber(int[] nums1, int[] nums2, int k) {
//         // Write your code here
//         if (k == 0)
//             return new int[0];

//         int m = nums1.length, n = nums2.length;
//         if (m + n < k) return null;
//         if (m + n == k) {
//             int[] results = merge(nums1, nums2, k);
//             return results;
//         } else {
//             int max = m >= k ? k : m;
//             int min = n >= k ? 0 : k - n;

//             int[] results = new int[k];
//             for(int i=0; i < k; ++i)
//                 results[i] = -0x7ffffff;
//             for(int i = min; i <= max; ++i) {
//                 int[] temp = merge(getMax(nums1, i), getMax(nums2, k - i), k);
//                 results = isGreater(results, 0, temp, 0) ? results : temp;
//             }
//             return results;
//         }
//     }

//     private int[] merge(int[] nums1, int[] nums2, int k) {
//         int[] results = new int[k];
//         if (k == 0) return results;
//         int i = 0, j = 0;
//         for(int l = 0; l < k; ++l) {
//             results[l] = isGreater(nums1, i, nums2, j) ? nums1[i++] : nums2[j++];
//         }
//         return results;
//     }

//     private boolean isGreater(int[] nums1, int i, int[] nums2, int j) {
//         for(; i < nums1.length && j < nums2.length; ++i, ++j) {
//             if (nums1[i] > nums2[j])
//                 return true;
//             if (nums1[i] < nums2[j])
//                 return false;
//         }
//         return i != nums1.length;
//     }

//     private int[] getMax(int[] nums, int k) {
//         if (k == 0)
//             return new int[0];
//         int[] results = new int[k];
//         int i = 0;
//         for(int j = 0; j < nums.length; ++j) {
//             while(nums.length - j + i > k && i > 0 && results[i-1] < nums[j])
//                 i--;
//             if (i < k)
//                 results[i++] = nums[j];
//         }
//         return results;
//     }
// }